File: go/storage/mkvs/urkel/node/node_test.go
package node

import (
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/oasislabs/ekiden/go/common/crypto/hash"
)

func TestSerializationLeafNode(t *testing.T) {
	key := Key("a golden key")
	var valueHash hash.Hash
	valueHash.FromBytes([]byte("value"))

	leafNode := &LeafNode{
		Round: 0xDEADBEEF,
		Key:   key,
		Value: &Value{
			Clean: true,
			Hash:  valueHash,
			Value: []byte("value"),
		},
	}

	rawLeafNodeFull, err := leafNode.MarshalBinary()
	require.NoError(t, err, "MarshalBinary")
	rawLeafNodeCompact, err := leafNode.CompactMarshalBinary()
	require.NoError(t, err, "CompactMarshalBinary")

	for _, rawLeafNode := range [][]byte{rawLeafNodeFull, rawLeafNodeCompact} {
		var decodedLeafNode LeafNode
		err = decodedLeafNode.UnmarshalBinary(rawLeafNode)
		require.NoError(t, err, "UnmarshalBinary")

		require.True(t, decodedLeafNode.Clean)
		require.Equal(t, leafNode.Round, decodedLeafNode.Round)
		require.Equal(t, leafNode.Key, decodedLeafNode.Key)
		require.True(t, decodedLeafNode.Value.Clean)
		require.Equal(t, leafNode.Value.Value, decodedLeafNode.Value.Value)
		require.NotNil(t, decodedLeafNode.Value.Value)
	}
}

func TestSerializationInternalNode(t *testing.T) {
	var valueHash hash.Hash
	valueHash.FromBytes([]byte("value"))
	var leafNode = &LeafNode{
		Key: []byte("a golden key"),
		Value: &Value{
			Clean: true,
			Hash:  valueHash,
			Value: []byte("value"),
		},
	}
	leafNode.UpdateHash()

	var leftHash hash.Hash
	leftHash.FromBytes([]byte("everyone move to the left"))
	var rightHash hash.Hash
	rightHash.FromBytes([]byte("everyone move to the right"))
	var label = Key("abc")
	var labelBitLength = Depth(24)

	intNode := &InternalNode{
		Round:          0xDEADBEEF,
		Label:          label,
		LabelBitLength: labelBitLength,
		LeafNode:       &Pointer{Clean: true, Node: leafNode, Hash: leafNode.Hash},
		Left:           &Pointer{Clean: true, Hash: leftHash},
		Right:          &Pointer{Clean: true, Hash: rightHash},
	}

	rawIntNodeFull, err := intNode.MarshalBinary()
	require.NoError(t, err, "MarshalBinary")
	rawIntNodeCompact, err := intNode.CompactMarshalBinary()
	require.NoError(t, err, "CompactMarshalBinary")

	for idx, rawIntNode := range [][]byte{rawIntNodeFull, rawIntNodeCompact} {
		var decodedIntNode InternalNode
		err = decodedIntNode.UnmarshalBinary(rawIntNode)
		require.NoError(t, err, "UnmarshalBinary")

		require.True(t, decodedIntNode.Clean)
		require.Equal(t, intNode.Round, decodedIntNode.Round)
		require.Equal(t, intNode.Label, decodedIntNode.Label)
		require.Equal(t, intNode.LabelBitLength, decodedIntNode.LabelBitLength)
		require.Equal(t, intNode.LeafNode.Hash, decodedIntNode.LeafNode.Hash)
		require.True(t, decodedIntNode.LeafNode.Clean)
		require.NotNil(t, decodedIntNode.LeafNode.Node)

		// Only check left/right for non-compact encoding.
		if idx == 0 {
			require.Equal(t, intNode.Left.Hash, decodedIntNode.Left.Hash)
			require.Equal(t, intNode.Right.Hash, decodedIntNode.Right.Hash)
			require.True(t, decodedIntNode.Left.Clean)
			require.True(t, decodedIntNode.Right.Clean)
			require.Nil(t, decodedIntNode.Left.Node)
			require.Nil(t, decodedIntNode.Right.Node)
		}
	}
}

func TestHashLeafNode(t *testing.T) {
	key := Key("a golden key")
	var valueHash hash.Hash
	valueHash.FromBytes([]byte("value"))

	leafNode := &LeafNode{
		Round: 0xDEADBEEF,
		Key:   key,
		Value: &Value{
			Clean: true,
			Hash:  valueHash,
			Value: []byte("value"),
		},
	}

	leafNode.UpdateHash()

	require.Equal(t, leafNode.Hash.String(), "7fc8d2e9142d15de712757dba87f6efd82a04a4ed1488e21ee95e3a7ec7a5fce")
}

func TestHashInternalNode(t *testing.T) {
	var leafNodeHash hash.Hash
	leafNodeHash.FromBytes([]byte("everyone stop here"))
	var leftHash hash.Hash
	leftHash.FromBytes([]byte("everyone move to the left"))
	var rightHash hash.Hash
	rightHash.FromBytes([]byte("everyone move to the right"))

	intNode := &InternalNode{
		Round:          0xDEADBEEF,
		Label:          Key("abc"),
		LabelBitLength: 23,
		LeafNode:       &Pointer{Clean: true, Hash: leafNodeHash},
		Left:           &Pointer{Clean: true, Hash: leftHash},
		Right:          &Pointer{Clean: true, Hash: rightHash},
	}

	intNode.UpdateHash()

	require.Equal(t, "e760353e9796f41b3bb2cfa2cf45f7e00ca687b6b84dc658e0ecadc906d5d21e", intNode.Hash.String())
}

func TestExtractLeafNode(t *testing.T) {
	key := Key("a golden key")
	var valueHash hash.Hash
	valueHash.FromBytes([]byte("value"))

	leafNode := &LeafNode{
		Clean: true,
		Round: 0xDEADBEEF,
		Key:   key,
		Value: &Value{
			Clean: true,
			Hash:  valueHash,
			Value: []byte("value"),
		},
	}

	exLeafNode := leafNode.Extract().(*LeafNode)

	require.False(t, leafNode == exLeafNode, "extracted node must have a different address")
	require.False(t, leafNode.Value == exLeafNode.Value, "extracted value must have a different address")
	require.Equal(t, true, exLeafNode.Clean, "extracted leaf must be clean")
	require.Equal(t, leafNode.Round, exLeafNode.Round, "extracted leaf must have the same round")
	require.Equal(t, key, exLeafNode.Key, "extracted leaf must have the same key")
	require.Equal(t, true, exLeafNode.Value.Clean, "extracted leaf must have clean value")
	require.Equal(t, valueHash, exLeafNode.Value.Hash, "extracted leaf's value must have the same hash")
	require.NotNil(t, exLeafNode.Value.Value, "extracted leaf's value must have non-nil value")
}

func TestExtractInternalNode(t *testing.T) {
	var leftHash hash.Hash
	leftHash.FromBytes([]byte("everyone move to the left"))
	var rightHash hash.Hash
	rightHash.FromBytes([]byte("everyone move to the right"))

	intNode := &InternalNode{
		Clean: true,
		Round: 0xDEADBEEF,
		Left:  &Pointer{Clean: true, Hash: leftHash},
		Right: &Pointer{Clean: true, Hash: rightHash},
	}

	exIntNode := intNode.Extract().(*InternalNode)

	require.False(t, intNode == exIntNode, "extracted node must have a different address")
	require.False(t, intNode.Left == exIntNode.Left, "extracted left pointer must have a different address")
	require.False(t, intNode.Right == exIntNode.Right, "extracted right pointer must have a different address")
	require.Equal(t, true, exIntNode.Clean, "extracted internal node must be clean")
	require.Equal(t, intNode.Round, exIntNode.Round, "extracted internal node must have the same round")
	require.Equal(t, leftHash, exIntNode.Left.Hash, "extracted left pointer must have the same hash")
	require.Equal(t, true, exIntNode.Left.Clean, "extracted left pointer must be clean")
	require.Equal(t, rightHash, exIntNode.Right.Hash, "extracted right pointer must have the same hash")
	require.Equal(t, true, exIntNode.Right.Clean, "extracted right pointer must be clean")
}

File: runtime/src/storage/mkvs/tree/node_test.rs
use std::{cell::RefCell, rc::Rc, str::FromStr};

use crate::{
    common::crypto::hash::Hash,
    storage::mkvs::{marshal::*, tree::*},
};

#[test]
fn test_serialization_leaf() {
    let leaf_node = LeafNode {
        key: b"a golden key".to_vec(),
        value: b"value".to_vec(),
        ..Default::default()
    };

    let marshaled = leaf_node.marshal_binary().expect("marshal");

    let mut decoded_leaf_node = LeafNode {
        ..Default::default()
    };
    decoded_leaf_node
        .unmarshal_binary(marshaled.as_slice())
        .expect("unmarshal");

    assert_eq!(true, decoded_leaf_node.clean);
    assert_eq!(leaf_node.key, decoded_leaf_node.key);
    assert_eq!(leaf_node.value, decoded_leaf_node.value);
}

#[test]
fn test_serialization_internal() {
    let mut leaf_node = LeafNode {
        key: b"a golden key".to_vec(),
        value: b"value".to_vec(),
        ..Default::default()
    };
    leaf_node.update_hash();
    let left_hash = Hash::digest_bytes(b"everyone move to the left");
    let right_hash = Hash::digest_bytes(b"everyone move to the right");

    let label: Key = b"abc".to_vec();
    let label_bit_length = 24 as Depth;

    let int_node = InternalNode {
        label: label,
        label_bit_length: label_bit_length,
        leaf_node: Rc::new(RefCell::new(NodePointer {
            clean: true,
            hash: leaf_node.get_hash(),
            node: Some(Rc::new(RefCell::new(NodeBox::Leaf(leaf_node)))),
            ..Default::default()
        })),
        left: Rc::new(RefCell::new(NodePointer {
            clean: true,
            hash: left_hash,
            ..Default::default()
        })),
        right: Rc::new(RefCell::new(NodePointer {
            clean: true,
            hash: right_hash,
            ..Default::default()
        })),
        ..Default::default()
    };

    let marshaled = int_node.marshal_binary().expect("marshal");

    let mut decoded_int_node = InternalNode {
        ..Default::default()
    };
    decoded_int_node
        .unmarshal_binary(marshaled.as_slice())
        .expect("unmarshal");

    assert_eq!(true, decoded_int_node.clean);
    assert_eq!(int_node.label, decoded_int_node.label);
    assert_eq!(int_node.label_bit_length, decoded_int_node.label_bit_length);
    assert_eq!(
        int_node.leaf_node.borrow().hash,
        decoded_int_node.leaf_node.borrow().hash
    );
    assert_eq!(
        int_node.left.borrow().hash,
        decoded_int_node.left.borrow().hash
    );
    assert_eq!(
        int_node.right.borrow().hash,
        decoded_int_node.right.borrow().hash
    );
    assert_eq!(true, decoded_int_node.leaf_node.borrow().clean);
    assert_eq!(true, decoded_int_node.left.borrow().clean);
    assert_eq!(true, decoded_int_node.right.borrow().clean);
    assert_eq!(true, decoded_int_node.leaf_node.borrow().node.is_some());
    assert_eq!(false, decoded_int_node.left.borrow().node.is_some());
    assert_eq!(false, decoded_int_node.right.borrow().node.is_some());
}

#[test]
fn test_hash_leaf() {
    let mut leaf_node = LeafNode {
        key: b"a golden key".to_vec(),
        value: b"value".to_vec(),
        ..Default::default()
    };

    leaf_node.update_hash();
    assert_eq!(
        leaf_node.hash,
        Hash::from_str("5c05183d4158b5920b16833acb78ccda464da83f720f824177b3a55a75f9fd88").unwrap()
    );
}

#[test]
fn test_hash_internal() {
    let leaf_node_hash = Hash::digest_bytes(b"everyone stop here");
    let left_hash = Hash::digest_bytes(b"everyone move to the left");
    let right_hash = Hash::digest_bytes(b"everyone move to the right");

    let mut int_node = InternalNode {
        label: b"abc".to_vec(),
        label_bit_length: 23,
        leaf_node: Rc::new(RefCell::new(NodePointer {
            clean: true,
            hash: leaf_node_hash,
            ..Default::default()
        })),
        left: Rc::new(RefCell::new(NodePointer {
            clean: true,
            hash: left_hash,
            ..Default::default()
        })),
        right: Rc::new(RefCell::new(NodePointer {
            clean: true,
            hash: right_hash,
            ..Default::default()
        })),
        ..Default::default()
    };

    int_node.update_hash();
    assert_eq!(
        int_node.hash,
        Hash::from_str("75c37c67c265e2c836f76dec35173fa336e976938ea46f088390a983e46efced").unwrap()
    );
}

#[test]
fn test_depth_type() {
    assert_eq! {0, (0 as Depth).to_bytes()};
    assert_eq! {2, (16 as Depth).to_bytes()};
    assert_eq! {3, (17 as Depth).to_bytes()};

    let mut dt: Depth = 0;
    assert_eq! {[0x0a as u8, 0x00].to_vec(), (10 as Depth).marshal_binary().unwrap()};
    dt.unmarshal_binary(&[0x0a, 0x00]).expect("unmarshal");
    assert_eq! {(10 as Depth), dt};
    assert_eq! {[0x04 as u8, 0x01].to_vec(), (260 as Depth).marshal_binary().unwrap()};
    dt.unmarshal_binary(&[0x04, 0x01]).expect("unmarshal");
    assert_eq! {(260 as Depth), dt};
}

#[test]
fn test_key_append_split_merge() {
    // append a single bit
    let key: Key = vec![0xf0];
    let new_key = key.append_bit(4, true);
    assert_eq!(vec![0xf8], new_key);
    let key: Key = vec![0xff];
    let new_key = key.append_bit(4, false);
    assert_eq!(vec![0xf7], new_key);
    let key: Key = vec![0xff];
    let new_key = key.append_bit(8, true);
    assert_eq!(vec![0xff, 0x80], new_key);
    let key: Key = vec![0xff];
    let new_key = key.append_bit(8, false);
    assert_eq!(vec![0xff, 0x00], new_key);

    // byte-aligned split
    let key: Key = vec![0xaa, 0xbb, 0xcc, 0xdd];
    let (p, s) = key.split(16, 32);
    assert_eq!(vec![0xaa, 0xbb], p);
    assert_eq!(vec![0xcc, 0xdd], s);

    // byte-aligned merge
    let key: Key = vec![0xaa, 0xbb];
    let new_key = key.merge(16, &vec![0xcc, 0xdd], 16);
    assert_eq!(vec![0xaa, 0xbb, 0xcc, 0xdd], new_key);

    // empty/full splits
    let key: Key = vec![0xaa, 0xbb, 0xcc, 0xdd];
    let (p, s) = key.split(0, 32);
    assert_eq!(Key::new(), p);
    assert_eq!(key, s);
    let (p, s) = key.split(32, 32);
    assert_eq!(key, p);
    assert_eq!(Key::new(), s);

    // empty merges
    let new_key = Key::new().merge(0, &vec![0xaa, 0xbb], 16);
    assert_eq!(vec![0xaa, 0xbb], new_key);
    let new_key: Key = vec![0xaa, 0xbb].merge(16, &Key::new(), 0);
    assert_eq!(vec![0xaa, 0xbb], new_key);

    // non byte-aligned split
    let key: Key = vec![0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef];
    let (p, s) = key.split(17, 64);
    assert_eq!(vec![0x01, 0x23, 0x00], p);
    assert_eq!(vec![0x8a, 0xcf, 0x13, 0x57, 0x9b, 0xde], s);

    // ...and merge
    let new_key = p.merge(17, &s, 64 - 17);
    assert_eq!(key, new_key);

    // non byte-aligned key length split.
    let key: Key = vec![0xff, 0xff, 0xff, 0xff];
    let (p, s) = key.split(21, 29);
    // Check that split cleans the last 3 unused bits!
    assert_eq!(vec![0xff, 0xff, 0xf8], p);
    assert_eq!(vec![0xff], s);

    // ...and merge
    let new_key = p.merge(21, &s, 8);
    // Merge doesn't obtain original key, because the split cleaned unused bits!
    assert_eq!(vec![0xff, 0xff, 0xff, 0xf8], new_key);

    // Special case with zero-length key.
    let key: Key = vec![0x80];
    let new_key = key.merge(0, &vec![0xf0], 4);
    assert_eq!(vec![0xf0], new_key);

    // Special case with extra bytes.
    let key: Key = vec![0x41, 0x6b, 0x00];
    let new_key = key.merge(16, &vec![0x37], 8);
    assert_eq!(vec![0x41, 0x6b, 0x37], new_key);
}

#[test]
fn test_key_common_prefix_len() {
    let key = Key::new();
    assert_eq!(0, key.common_prefix_len(0, &Key::new(), 0));

    let key: Key = vec![0xff, 0xff];
    assert_eq!(16, key.common_prefix_len(16, &vec![0xff, 0xff, 0xff], 24));

    let key: Key = vec![0xff, 0xff, 0xff];
    assert_eq!(16, key.common_prefix_len(24, &vec![0xff, 0xff], 16));

    let key: Key = vec![0xff, 0xff, 0xff];
    assert_eq!(24, key.common_prefix_len(24, &vec![0xff, 0xff, 0xff], 24));

    let key: Key = vec![0xab, 0xcd, 0xef, 0xff];
    assert_eq!(
        23,
        key.common_prefix_len(32, &vec![0xab, 0xcd, 0xee, 0xff], 32)
    );

    let key: Key = vec![0xab, 0xcd];
    assert_eq!(12, key.common_prefix_len(13, &vec![0xab, 0xcd], 12));
    assert_eq!(12, key.common_prefix_len(12, &vec![0xab, 0xcd], 13));
}
